docs/proposals/ose-rebase.md

```diff
@@ -0,0 +1,120 @@
+moving all the changes in OSE that diverge from Origin into a set of commits that will sit
+on top of the Origin code[0] and a Jenkins job that will be rebasing master on a daily basis.
+Rebases are going to continue to happen only for the master branch.
+
+There are two types of commits that need to be carried on top of Origin:
+1. long-term carries related to branding and release tooling (master and release branches)
+2. CVEs and other backported bugfixes (release branches only)
+
+The Origin repository is set as a remote `upstream` branch in OSE so in master we can
+simply rebase upstream/master. Before the actual rebase happens, the last commit produced
+by tito during the previous rebase should be squashed into the main tito commit (one
+of the [CARRY] commits). If there is an intermediate commit that bumps the web console
+it should be dropped and a new updated may be added further down the build process. The
+tag that points to the last tito commit needs to be retaged because the squash is dropping
+it. If any conflicts are hit, manual resolution will be required as it is already the case
+with `git merge`. In the future, we can extend the conflict resolution mechanism in Jenkins
+to send an e-mail to the author and committer of the carry commits.
+
+A nightly build of OSE should follow the steps below in order to rebase on top of Origin:
+```sh
+git clone git@github.com/openshift/ose.git
+pushd ose
+git remote add upstream git@github.com/openshift/origin.git
+git fetch upstream
+git checkout master
+
+# Note down the current latest tag because squashing the tito commit into specfile updates
+# will drop it and tito tag needs it further down the build process in order to determine
+# the next (new) tag.
+last_tag="$( git describe --abbrev=0 --tags )"
+
+# sanity check that ensures the latest commits are in the right order.
+#
+# Either the last two commits are the tito tag commit and the specfile updates (where every
+# tito commit is eventually squashed):
+#
+#         .-Other [CARRY] commit
+#        /  .-Other [CARRY] commit
+#       /  /  .-[CARRY] Specfile updates
+#      /  /  /  .-Tito tag commit
+#     /  /  /  /
+# m--c--c--s--t
+#  \           \
+#   \           `-ose/master/HEAD,v3.5.0.23
+#    `-origin/master/HEAD
+#
+# or the last three commits are the tito tag commit, a web console bump that got created
+# during the previous rebase, and the specfile updates:
+#
+#         .-Other [CARRY] commit
+#        /  .-[CARRY] Specfile updates
+#       /  /  .-[DROP] webconsole bump
+#      /  /  /  .-Tito tag commit
+#     /  /  /  /
+# m--c--c--s--t
+#  \           \
+#   \           `-ose/master/HEAD,v3.5.0.23
+#    `-origin/master/HEAD
+#
+# In the first case, we simply want to reset the last two commits and commit them as one.
+# In the second case, we want the same thing as in the first case plus to completely drop
+# the webconsole bump since a new one may be created during the current build.
```

Comment:
@tdawson this is not related to the branding commit (do we have a script that generates the branding diff?)

When we merge Origin master into OSE master, hack/vendor.sh is rolling back the console from console/master (what is vendored inside Origin master) to console/enterprise-3.x (where x is the new release for which no branch has been cut yet).

When a release approaches and we cut the new branch and switch from master to enterprise-3.x rebases, we merge Origin release-1.x into OSE enterprise-3.x, hack/vendor.sh may or may not produce any additional diff for the console.

For example:
https://github.com/openshift/ose/commits/enterprise-3.5
Automatic commit of package [atomic-openshift] release [3.5.0.35-1]. 
Automatic commit of package [atomic-openshift] release [3.5.0.33-1]. 

and others have no web console bump followed by them, most likely because there was no diff between the vendored console in Origin and OSE.

Another thing that I just realized is that when we merge Origin master to OSE master, we also vendor the web console **master** even though we are targetting the enterprise branch.
https://github.com/openshift/aos-cd-jobs/blob/626a5f7f31a6d0da8657ed175da603a7aa04e76c/scripts/merge-and-build.sh#L69